File: frontend/lib/finder.ts
import { fetchWithTimeout } from "./http";

export interface FoundBusiness {
  businessName: string;
  address: string;
  website: string;
  phone: string;
  lat: number | null;
  lng: number | null;
  source: "openstreetmap" | "sample";
}

// Map common niche words to OpenStreetMap tag filters.
const NICHE_TAGS: Record<string, string[]> = {
  dentist: ['["amenity"="dentist"]', '["healthcare"="dentist"]'],
  restaurant: ['["amenity"="restaurant"]'],
  cafe: ['["amenity"="cafe"]'],
  bar: ['["amenity"="bar"]', '["amenity"="pub"]'],
  plumber: ['["craft"="plumber"]', '["shop"="trade"]'],
  electrician: ['["craft"="electrician"]'],
  lawyer: ['["office"="lawyer"]', '["amenity"="lawyer"]'],
  accountant: ['["office"="accountant"]'],
  hairdresser: ['["shop"="hairdresser"]'],
  salon: ['["shop"="hairdresser"]', '["shop"="beauty"]'],
  gym: ['["leisure"="fitness_centre"]', '["amenity"="gym"]'],
  bakery: ['["shop"="bakery"]'],
  florist: ['["shop"="florist"]'],
  mechanic: ['["shop"="car_repair"]'],
  roofer: ['["craft"="roofer"]'],
  builder: ['["craft"="builder"]', '["office"="construction_company"]'],
  photographer: ['["craft"="photographer"]', '["shop"="photo"]'],
  veterinary: ['["amenity"="veterinary"]'],
  optician: ['["shop"="optician"]'],
  pharmacy: ['["amenity"="pharmacy"]'],
  hotel: ['["tourism"="hotel"]'],
  realestate: ['["office"="estate_agent"]'],
};

function tagsForNiche(niche: string): string[] {
  const key = niche.toLowerCase().replace(/[^a-z]/g, "");
  for (const k of Object.keys(NICHE_TAGS)) {
    if (key.includes(k)) return NICHE_TAGS[k];
  }
  // Generic fallback: fuzzy name match across shops & offices.
  const safe = niche.replace(/["\\]/g, "");
  return [
    `["shop"]["name"~"${safe}",i]`,
    `["office"]["name"~"${safe}",i]`,
    `["amenity"]["name"~"${safe}",i]`,
  ];
}

async function geocodeCity(city: string): Promise<{ lat: number; lng: number } | null> {
  try {
    const res = await fetchWithTimeout(
      `https://nominatim.openstreetmap.org/search?format=json&limit=1&q=${encodeURIComponent(city)}`,
      { headers: { accept: "application/json" }, timeoutMs: 8000 }
    );
    const data = (await res.json()) as Array<{ lat: string; lon: string }>;
    if (!data.length) return null;
    return { lat: parseFloat(data[0].lat), lng: parseFloat(data[0].lon) };
  } catch {
    return null;
  }
}

export async function findBusinesses(
  niche: string,
  city: string,
  limit = 20
): Promise<{ results: FoundBusiness[]; source: "openstreetmap" | "sample" }> {
  const geo = await geocodeCity(city);
  if (geo) {
    try {
      const radius = 12000; // metres
      const filters = tagsForNiche(niche)
        .map(
          (t) =>
            `node${t}(around:${radius},${geo.lat},${geo.lng});way${t}(around:${radius},${geo.lat},${geo.lng});`
        )
        .join("");
      const query = `[out:json][timeout:20];(${filters});out center ${limit};`;
      const res = await fetchWithTimeout("https://overpass-api.de/api/interpreter", {
        method: "POST",
        headers: { "content-type": "application/x-www-form-urlencoded" },
        body: "data=" + encodeURIComponent(query),
        timeoutMs: 20000,
      });
      const data = (await res.json()) as { elements: any[] };
      const results = (data.elements || [])
        .map((el) => mapElement(el))
        .filter((b): b is FoundBusiness => b !== null)
        .slice(0, limit);
      if (results.length) return { results, source: "openstreetmap" };
    } catch {
      // fall through to sample
    }
  }
  return { results: sampleBusinesses(niche, city, limit), source: "sample" };
}

function mapElement(el: any): FoundBusiness | null {
  const t = el.tags || {};
  const name = t.name || t["name:en"];
  if (!name) return null;
  const addr = [t["addr:housenumber"], t["addr:street"], t["addr:city"], t["addr:postcode"]]
    .filter(Boolean)
    .join(", ");
  const lat = el.lat ?? el.center?.lat ?? null;
  const lng = el.lon ?? el.center?.lon ?? null;
  return {
    businessName: name,
    address: addr,
    website: (t.website || t["contact:website"] || t.url || "").trim(),
    phone: (t.phone || t["contact:phone"] || "").trim(),
    lat,
    lng,
    source: "openstreetmap",
  };
}

// Deterministic sample generator so the finder always returns believable
// results when open-map data is unavailable (offline/sandboxed environments).
function sampleBusinesses(niche: string, city: string, limit: number): FoundBusiness[] {
  const nicheClean = niche.trim() || "Local business";
  const cityClean = city.trim() || "Town";
  const prefixes = ["Sunrise", "Elm Street", "Premier", "Harbour", "Oakwood", "Cornerstone", "Bright", "Maple", "Riverside", "Golden", "Summit", "Ironclad", "Blue Sky", "Evergreen", "Downtown", "Heritage", "Anchor", "Crestview", "Willow", "Pioneer"];
  const suffixes = ["Co.", "& Sons", "Group", "Studio", "Services", "Ltd", "Practice", "Collective", "Partners", "Works"];
  // Some deliberately have no / weak websites — these become the best leads.
  const out: FoundBusiness[] = [];
  for (let i = 0; i < Math.min(limit, prefixes.length); i++) {
    const hasSite = i % 3 !== 0; // ~1/3 have no website at all
    const slug = prefixes[i].toLowerCase().replace(/[^a-z]/g, "");
    out.push({
      businessName: `${prefixes[i]} ${titleCase(nicheClean)} ${suffixes[i % suffixes.length]}`,
      address: `${100 + i * 7} ${prefixes[i]} Road, ${cityClean}`,
      website: hasSite ? `http://${slug}${slug.length % 2 ? "" : "-"}${cityClean.toLowerCase().replace(/[^a-z]/g, "")}.com` : "",
      phone: `+1 555 0${(100 + i).toString().slice(-3)} ${(2000 + i * 13).toString().slice(-4)}`,
      lat: null,
      lng: null,
      source: "sample",
    });
  }
  return out;
}

function titleCase(s: string) {
  return s.replace(/\b\w/g, (c) => c.toUpperCase());
}

File: backend/app/services/finder.py
"""Business finder.

Provider order: SerpApi (if key set) → OpenStreetMap (Nominatim + Overpass) →
deterministic sample data, so the finder always returns results.
"""
from __future__ import annotations

import httpx

from ..config import settings
from ..core.security import domain_of

NICHE_TAGS: dict[str, list[str]] = {
    "dentist": ['["amenity"="dentist"]'],
    "restaurant": ['["amenity"="restaurant"]'],
    "cafe": ['["amenity"="cafe"]'],
    "plumber": ['["craft"="plumber"]'],
    "dakwerker": ['["craft"="roofer"]'],
    "roofer": ['["craft"="roofer"]'],
    "electrician": ['["craft"="electrician"]'],
    "lawyer": ['["office"="lawyer"]'],
    "hairdresser": ['["shop"="hairdresser"]'],
    "gym": ['["leisure"="fitness_centre"]'],
    "bakery": ['["shop"="bakery"]'],
    "garage": ['["shop"="car_repair"]'],
    "mechanic": ['["shop"="car_repair"]'],
    "detailing": ['["shop"="car_repair"]', '["shop"="car"]'],
    "hvac": ['["craft"="hvac"]'],
    "solar": ['["craft"="electrician"]'],
    "landscaper": ['["landuse"="garden_centre"]', '["shop"="garden_centre"]'],
    "hotel": ['["tourism"="hotel"]'],
}


def _client() -> httpx.Client:
    return httpx.Client(
        timeout=settings.request_timeout_s + 6,
        headers={"user-agent": "LeadlyBot/1.0 (+https://getleadly.net)"},
        proxy=settings.outbound_proxy or None,
    )


def _tags(niche: str) -> list[str]:
    key = "".join(c for c in niche.lower() if c.isalpha())
    for k, v in NICHE_TAGS.items():
        if k in key:
            return v
    safe = niche.replace('"', "").replace("\\", "")
    return [f'["shop"]["name"~"{safe}",i]', f'["office"]["name"~"{safe}",i]', f'["amenity"]["name"~"{safe}",i]']


def find_businesses(niche: str, city: str, limit: int = 20) -> dict:
    if settings.serpapi_key:
        try:
            return {"source": "serpapi", "results": _serpapi(niche, city, limit)}
        except Exception:
            pass
    try:
        results = _openstreetmap(niche, city, limit)
        if results:
            return {"source": "openstreetmap", "results": results}
    except Exception:
        pass
    return {"source": "sample", "results": _sample(niche, city, limit)}


def _serpapi(niche: str, city: str, limit: int) -> list[dict]:
    with _client() as c:
        r = c.get("https://serpapi.com/search.json", params={
            "engine": "google_maps", "type": "search",
            "q": f"{niche} in {city}", "api_key": settings.serpapi_key,
        })
        data = r.json()
    out = []
    for it in (data.get("local_results") or [])[:limit]:
        site = it.get("website", "")
        out.append({
            "business_name": it.get("title", ""),
            "website": site, "domain": domain_of(site),
            "phone": it.get("phone", ""), "address": it.get("address", ""),
            "city": city, "country": "",
            "lat": (it.get("gps_coordinates") or {}).get("latitude"),
            "lng": (it.get("gps_coordinates") or {}).get("longitude"),
            "category": it.get("type", ""),
            "google_rating": it.get("rating"), "review_count": it.get("reviews"),
            "maps_url": it.get("place_id_search", ""),
            "source": "serpapi", "source_query": f"{niche} in {city}",
        })
    return out


def _openstreetmap(niche: str, city: str, limit: int) -> list[dict]:
    with _client() as c:
        geo = c.get("https://nominatim.openstreetmap.org/search",
                    params={"format": "json", "limit": 1, "q": city}).json()
        if not geo:
            return []
        lat, lng = float(geo[0]["lat"]), float(geo[0]["lon"])
        filters = "".join(
            f"node{t}(around:12000,{lat},{lng});way{t}(around:12000,{lat},{lng});"
            for t in _tags(niche)
        )
        query = f"[out:json][timeout:20];({filters});out center {limit};"
        data = c.post("https://overpass-api.de/api/interpreter",
                      content="data=" + query,
                      headers={"content-type": "application/x-www-form-urlencoded"}).json()
    out = []
    for el in data.get("elements", [])[:limit]:
        t = el.get("tags", {})
        name = t.get("name") or t.get("name:en")
        if not name:
            continue
        site = (t.get("website") or t.get("contact:website") or "").strip()
        out.append({
            "business_name": name,
            "website": site, "domain": domain_of(site),
            "phone": (t.get("phone") or t.get("contact:phone") or "").strip(),
            "address": ", ".join(filter(None, [t.get("addr:street"), t.get("addr:city")])),
            "city": city, "country": "",
            "lat": el.get("lat") or (el.get("center") or {}).get("lat"),
            "lng": el.get("lon") or (el.get("center") or {}).get("lon"),
            "category": niche, "google_rating": None, "review_count": None,
            "maps_url": "", "source": "openstreetmap", "source_query": f"{niche} in {city}",
        })
    return out


def _sample(niche: str, city: str, limit: int) -> list[dict]:
    prefixes = ["Sunrise", "Elm Street", "Premier", "Harbour", "Oakwood", "Cornerstone",
                "Bright", "Maple", "Riverside", "Golden", "Summit", "Ironclad",
                "Blue Sky", "Evergreen", "Downtown", "Heritage", "Anchor", "Crestview"]
    out = []
    for i in range(min(limit, len(prefixes))):
        has_site = i % 3 != 0
        slug = "".join(c for c in prefixes[i].lower() if c.isalpha())
        citys = "".join(c for c in city.lower() if c.isalpha())
        site = f"http://{slug}{citys}.com" if has_site else ""
        out.append({
            "business_name": f"{prefixes[i]} {niche.title()}",
            "website": site, "domain": domain_of(site),
            "phone": f"+1 555 0{100+i:03d}", "address": f"{100+i*7} {prefixes[i]} Rd, {city}",
            "city": city, "country": "",
            "lat": None, "lng": None, "category": niche,
            "google_rating": round(3.5 + (i % 5) * 0.3, 1), "review_count": 5 + i * 7,
            "maps_url": "", "source": "sample", "source_query": f"{niche} in {city}",
        })
    return out
